core/src/splitting/factory/InstrumentalSplittingRuleFactory.cpp

```diff
@@ -23,7 +23,7 @@ namespace grf {
 std::unique_ptr<SplittingRule> InstrumentalSplittingRuleFactory::create(const Data& data,
                                                                         const TreeOptions& options) const {
   return std::unique_ptr<SplittingRule>(new InstrumentalSplittingRule(
-      data.get_max_num_unique_values(),
+      data.get_num_rows(),
```

Comment:
That sounds very worthwhile, do you have design thoughts how to put that in place?

Right before the splitting rule factory is called we know the sub sample sample size (the length of` nodes[0]` in the previous line):

https://github.com/grf-labs/grf/blob/master/core/src/tree/TreeTrainer.cpp#L60

We can move the scratch arrays initialization to the Factory? (I can do that in a follow up PR)

Reply:
Could we change the signature of `SplittingRuleFactory` to take in the subsample size instead of `Data`? And sounds good to discuss and make the change in a follow-up PR.

Reply:
👍